### ISSUE TITLE
torch.compile should not recompiles when `.requires_grad=True` under `torch.no_grad()` context

### ISSUE THREAD

**Issue by gau-nernst** [pytorch/pytorch#131975]
### 🐛 Describe the bug

```python
import torch

@torch.compile()
def f(a, b):
    return (a + b).clip(0)

a = torch.randn(1024, 1024)
b = torch.randn(1024, 1024)

with torch.no_grad():
    f(a, b)
    a.requires_grad_(True)
    f(a, b)
```

Run with `TORCH_LOGS="recompiles"`

```
V0727 08:08:44.508000 30548 torch/_dynamo/guards.py:2688] [0/1] [__recompiles] Recompiling function f in /teamspace/studios/this_studio/debug.py:3
V0727 08:08:44.508000 30548 torch/_dynamo/guards.py:2688] [0/1] [__recompiles]     triggered by the following guard failure(s):
V0727 08:08:44.508000 30548 torch/_dynamo/guards.py:2688] [0/1] [__recompiles]     - 0/0: tensor 'L['a']' requires_grad mismatch. expected requires_grad=0
```

Putting `torch.no_grad()` inside the compiled function also yields the same results. Although this is not a bug per se, minimize recompiles should be beneficial, especially to avoid hitting recompiles limit. From my understanding, there shouldn't be any correctness problem if we don't recompile, since `.requires_grad` is irrelevant inside the `torch.no_grad()` context.

### Versions

Latest torch nightly 2.5.0.dev20240726

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames

**Comment by vulkomilev** [comment]
Hi I am not very sure but I think requires_grad_ will overwrite the behavior of torch.no_grad()

**Comment by yf225** [comment]
I think `requires_grad_` doesn't overwrite the behavior of `torch.no_grad()` in eager mode:
```python
import torch

def f(a, b):
    return torch.matmul(a, b)

a = torch. randn(1024, 1024)
b = torch. randn(1024, 1024, requires_grad=True)

with torch.no_grad():
    f(a, b)
    a.requires_grad_(True)
    out = f(a, b)
    out.sum().backward()    # RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
    print(a.grad)
```
yeah it would be ideal to avoid the recompile in compile mode.

**Comment by Xynonners** [comment]
still seems to happen

**Comment by zou3519** [comment]
The problem is that we don't know if the code branches on requires_grad=True or not. We should be able to detect for the user code but idk for the c++ PyTorch code (or if that even matters)

**Comment by StrongerXi** [comment]
@zou3519 is right. More concretely, right now we preemptively guard on `requires_grad` of tensors, as part of the `TensorCheck` guard (which also preemptively guard on other properties like dispatch key, etc.): 
https://github.com/pytorch/pytorch/blob/3849fd13de1a5ee727fd64351edb0d3b1bb637a9/torch/csrc/dynamo/guards.cpp#L207-L213

But in theory, we only need to guard on a property if the compilation specialized its behavior based on the property. In this case it's either:
1. User code directly uses the `requires_grad` attribute, e.g., to do conditional branch, which affects the code path Dynamo is tracing.
2. AOTDispatcher uses it for AOTAutograd (it stays faithful to `torch.no_grad` as well) or some other C++ PyTorch code, which affects the output of the forward and backward graphs.

In this issue, the preemptive guard basically diverged from what (2) would've guarded on, given its knowledge and usage of `requires_grad` and `torch.no_grad`. So to fix this properly, I think we'd need to remove the `requires_grad` check from `TensorCheck` guard, and lazily emit `requires_grad` check during (2) to resolve this issue, and during (1) to preserve complete soundness. This will also fix #139476, but similarly could also increase guard latency.

@gau-nernst would you mind sharing how you run into this? Was this from some model/framework?

@bdhirsh @anijain2305 wdyt? Is this worth fixing?

**Comment by bdhirsh** [comment]
As richard mentioned, there is technically a risk that if you have some C++ code that we trace (say, a custom op meta/decomp written in C++) that branches on the input requires_grad field for whatever reason), we will not know to properly recompile. This seems like one of those "hard to imagine why someone would do this, but is technically a source of silent correctness" cases.

My tentative take would be - if we already have an API for allowing the user to remove guards (and risking their own correctness problems for better perf / fewer recompiles) - we may as well tell users to reach for that tool here?

**Comment by StrongerXi** [comment]
> if we already have an API for allowing the user to remove guards (and risking their own correctness problems for better perf / fewer recompiles) - we may as well tell users to reach for that tool here?

Yeah I think that's the best workaround for now, this feels like a niche enough issue to warrant a niche solution, and conveniently we just added such an API:) #150936. Example:
```python
        def guard_filter_fn(entries):
            return [
                not (entry.name == "y" and entry.value is None) for entry in entries
            ]

        @torch.compile(fullgraph=True, options={"guard_filter_fn": guard_filter_fn})
        def fn(x, y):
            if y is not None:
                x += y
            return x
```

l'll close this issue for now.

**Comment by gau-nernst** [comment]
@StrongerXi It's been a long time since I opened this issue (nearly 1 year ago!), so I don't rmb the exact problem / how I ran into this. Just think that it's a nice feature to have to avoid recompilation, which is especially important when:
1. Compilation is expensive / long
2. Avoid falling back to eager due to hitting recompile limit (can be avoided by raising recompile limit, but not too ideal)

Understand it's not as simple as it seems. Thank you for taking the time looking into it.
